### PR TITLE
Swift integration

### PR DESCRIPTION
**Pull request by jasondebolt**
Swift integration for Codewars.

Allows Swift to be used as a programming language for interactive code challenges on Codewars.com

Includes:
      - Updated Makefile and docker-compose.yml files to build a "swift-runner" docker image.
      - "swift.yml" file with both cw-2 and xctest examples.
      - "PrintObvserver.swift" file with Codewars specific test suite stdout format.
      - "codewars-build-presets.ini" file to describe how to build Swift from source for Codewars.
      - "swift.js" file to allow both cw-2 and xctest code to be run from the CLI.
      - "swift_spec.js" file with test cases for both cw-2 and xctest.

### CODE REVIEW

**Comment by OverZealous** [comment]
This is awesome! I can't wait to see Swift on CW.

File: frameworks/swift/xctest/PrintObserver.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      }
+      return testCase.name
+    }
+
+    func testCaseWillStart(_ testCase: XCTestCase) {
+        printAndFlush("<IT::>\(getMethodName(testCase))")
+    }
+
+    func testCase(_ testCase: XCTestCase, didFailWithDescription description: String, inFile filePath: String?, atLine lineNumber: UInt) {
+        printAndFlush("<ERROR::>\(getMethodName(testCase)) : \(description)")
```
Comment (OverZealous) [inline frameworks/swift/xctest/PrintObserver.swift]: One thing I'm not seeing here is making sure to replace `\n` with `<:LF:>`.

Without this, any multi-line content will get picked up as normal `stdout` content.

Comment (OverZealous) [inline frameworks/swift/xctest/PrintObserver.swift]: It needs to replace every newline—I'm very new to Swift, so I don't know if that's how `terminator` works or not.

What you want to make sure is something like this:
```
<FAILED::>error on line 1 of foo
on line 2 of bar
```

Is rendered like this:

```
<FAILED::>error on line 1 of foo<:LF:>on line 2 of bar
```

This helps the runner to keep related blocks of text together.

Comment (OverZealous) [inline frameworks/swift/xctest/PrintObserver.swift]: OK, I just tested it in a Swift playground, that appears to only change the end terminator, which is sort of the exact opposite behavior you want.

Comment (jasondebolt) [inline frameworks/swift/xctest/PrintObserver.swift]: Ahh, I see. I'll update the tests and do something similar to the "format" function in frameworks/cpp/main.cpp.

**Comment by jhoffner** [comment]
Awesome work!

**Comment by jasondebolt** [comment]
Thanks! Let me what changes need to be made. This is my first codewars pull
request. I have no idea if I'm doing this correctly.

**Comment by jhoffner** [comment]
Will do. I'm having issues building it at the moment but it seems to possibly be an issue with my Docker environment. Will let you know if I am unable to get passed it.

**Comment by jasondebolt** [comment]
Ok. Unfortunately, it can take a very long time to build.  It takes 90
minutes to build the whole container on my MacBook. Feel free to send me
any errors so I can help you diagnose quickly. I'll also be available all
day tomorrow if you need anything.

Jason

**Comment by jasondebolt** [comment]
Hi Jake,

Were you able resolve the Docker issue and build the Swift container? Let
me know what else needs to be done. I'm happy to help.

- Jason

**Comment by jhoffner** [comment]
Yep I did. It took a while to push that bad boy up to docker hub. I still need to test, which I will get into tomorrow. Got caught up with other stuff today. I've got it running locally and from the little I tested it looks great! One thing is that ideally it would be nice if each individual assertion resulted in a `Test Passed` message, just so users could see how many assertions were being ran, but not really a big deal (and not always possible). 

I'm going to work on getting it pushed to preview tomorrow. The image is so big that I'll need to go about upgrading in a slower way than normal so I can't quickly do it right now.

**Comment by jasondebolt** [comment]
Awesome.

The assertion thing should be possible. Take a look at the following two
files:

https://github.com/apple/swift-corelibs-xctest/blob/master/Sources/XCTest/Public/XCTAssert.swift
https://github.com/apple/swift-corelibs-xctest/blob/master/Sources/XCTest/Private/PrintObserver.swift

We can keep the <DESCRIBE::>, <IT::>, <ERROR::>, <FAILURE::>, and
<COMPLETEDIN::> in the "PrintObvserver.swift" file and add the <PASSED::>,
<FAILED::>, and <ERROR:> commands to the XCTAssert.swift file (specifically
the in the _XCTEvaulateAssertion.swift file). It probably makes sense to
remove the <PASSED::> command at the test method level if we are adding
<PASSED::> at the assertion level. Let me know what makes the most sense
here.


private func _XCTEvaluateAssertion(_ assertion: _XCTAssertion, message:
@autoclosure () -> String = "", file: StaticString = #file, line: UInt =
#line, expression: () throws -> _XCTAssertionResult) {
    let result: _XCTAssertionResult
    do {
        result = try expression()
    } catch {
        result = .unexpectedFailure(error)
    }

    switch result {
    case .success:
        *print("<PASSED::>")*
        return
    default:
        if let currentTestCase = XCTCurrentTestCase {
*            // `true` if the failure being reported was the result of a
failed assertion*
*            // `false` if it was the result of an uncaught exception.*
*            if result.isExpected {*
*              print("<FAILED::>\(result.failureDescription(assertion)) -
\(message())")*
*            } else {*
*              print("<ERROR::>\(result.failureDescription(assertion)) -
\(message())")*
*            }*
            currentTestCase.recordFailure(
                withDescription: "\(result.failureDescription(assertion)) -
\(message())",
                inFile: String(describing: file),
                atLine: line,
                expected: result.isExpected)
        }
    }
}

File: frameworks/swift/xctest/PrintObserver.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    private lazy var dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm:ss.SSS"
+        return formatter
+    }()
+
+    fileprivate func printAndFlush(_ message: String) {
+        print(message)
+        fflush(stdout)
```
Comment (jhoffner) [inline frameworks/swift/xctest/PrintObserver.swift]: should this `fflush` have two f's?

**Comment by jasondebolt** [comment]
Yes.

**Comment by jhoffner** [comment]
@jasondebolt a quick update. There are some "thin-pool" issues with Docker that are happening both on my local and production, where I guess there are so many images loaded that Docker thinks it is out of space - even though it shouldn't be. Basically until I can get this fixed there is a wall that doesn't allow me to add anymore images, so I'm not able to deploy the Swift image yet. I've spent almost 2 days on this at this point and its kind of been a mess. I need to take a break from it so I'm going to pick it back up on Monday. I expect to have Swift out on CW sometime early next week.

**Comment by jasondebolt** [comment]
Sorry to hear that. You're probably a Docker ninja, so I'm sure you'll
figure it out. Let me know if there's anything I can do to help.

Have you tried checking your partitions for capacity? You probably already
know these commands, but I like to use "$df -h" and "$du -sch ./* | sort
-n" when docker starts telling me I'm running out of disk space. I also
like to use the following commands to clean up my docker environment:

$ docker stop $(docker ps -a -q) --> Stops all docker containers
$ docker rm $(docker ps -a -q) --> Removes all docker containers
$ docker rmi $(docker images | grep "^<none>" | awk "{print $3}") -->
Removes all untagged images.

However, even after running the above commands, docker can still hog quite
a bit of disk space. Is this case, I've had to remove all Docker containers
from the VM on my mac by deleting the following file:

/Users/jasondebolt/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/Docker.qcow2

The swift runner image is very large. I just realized that I'm not deleting
the Swift source code and libraries from which Swift is built. Removing the
source code could save up to 4 GB of space. I'm going to try adding "RUN rm
-rf $SWIFT_SOURCE_DIR/*" in the swift.docker file and rebuild the image.
That should significantly cut down on the image size. If that works, I'll
create a pull request if that's ok with you.


- Jason

**Comment by jhoffner** [comment]
Thanks @jasondebolt. Yeah unfortunately it wasn't an issue with disc space on the machine, it actually had 15gb free. For some reason the Docker thin-pool wasn't expanding the way that it should. 

Having the image be 4GB lighter would be a huge help! I still need to fix the thin-pool issue regardless but that would help me get Swift out there sooner, plus its just a lot easier to manage.

**Comment by jasondebolt** [comment]
Cool. I just created another pull request which should reduce image size by
~4GB and enables individual test assertion output.

https://github.com/Codewars/codewars-runner-cli/pull/242

Have a great weekend!

**Comment by g964** [comment]
Copy of Swift translation #830

Sorry to pollute with a long post... I translated to swift <https://www.codewars.com/kata/5726f813c8dcebf5ed000a6b>
I got lots of
```
Time: 4898ms Passed: 13 Failed: 0
Test Results:
  SolutionTest
  countKprimes (9 Passed, 0 Failed)
  puzzle (4 Passed, 0 Failed)
  Random Examples
STDERR:
0  swift           0x00000000034bf568 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift           0x00000000034c0146
2  libpthread.so.0 0x00007f696139e330
3  libpthread.so.0 0x00007f6961714801
4  libpthread.so.0 0x00007f69617153ff
5  libpthread.so.0 0x00007f6961715d50
6  libpthread.so.0 0x00007f6961714b2e
7  libpthread.so.0 0x00007f6961715e75
8  libXCTest.so    0x00007f695c64e6ab
9  libXCTest.so    0x00007f695c6251a2
10 libXCTest.so    0x00007f695c64e660
11 libXCTest.so    0x00007f695c64e3e0
12 libXCTest.so    0x00007f695c624280 _TFC6XCTest10XCTestCase10invokeTestfT_T_ + 112
13 libXCTest.so    0x00007f695c63c219 _TTSf4g_n___TFC6XCTest10XCTestCase7performfCS_9XCTestRunT_ + 313
14 libXCTest.so    0x00007f695c62414e _TFC6XCTest10XCTestCase7performfCS_9XCTestRunT_ + 14
15 libXCTest.so    0x00007f695c623f45 _TFC6XCTest6XCTest3runfT_T_ + 629
16 libXCTest.so    0x00007f695c63cee5 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 789
17 libXCTest.so    0x00007f695c63d3b3 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 2019
18 libXCTest.so    0x00007f695c63d3b3 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 2019
19 libXCTest.so    0x00007f695c6296db _TF6XCTest7XCTMainFGSaT13testCaseClassMCS_10XCTestCase8allTestsGSaTSSFzS0_T_____Os5Never + 4283
20 libXCTest.so    0x00007f69617140e9 _TF6XCTest7XCTMainFGSaT13testCaseClassMCS_10XCTestCase8allTestsGSaTSSFzS0_T_____Os5Never + 84851401
21 swift           0x0000000000c795f2 llvm::MCJIT::runFunction(llvm::Function*, llvm::ArrayRef) + 1042
22 swift           0x0000000000c7d0ff llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::vector > const&, char const* const*) + 1215
23 swift           0x000000000083ca6f swift::RunImmediately(swift::CompilerInstance&, std::vector > const&, swift::IRGenOptions&, swift::SILOptions const&) + 2303
24 swift           0x00000000008364f8
25 swift           0x0000000000831377 swift::performFrontend(llvm::ArrayRef, char const*, void*, swift::FrontendObserver*) + 2551
26 swift           0x00000000007f8788 main + 3016
27 libc.so.6       0x00007f695fb47f45 __libc_start_main + 245
28 swift           0x00000000007f5e56
Stack dump:
0.	Program arguments: /swift-vol/swift-gen/usr/bin/swift -frontend -interpret /tmp/swift117117-18-p227nr.n1qsuzyqfr/solution.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -module-name solution
````
I don't understand why that:-( Is there a problem with Swift and if yes, which problem? These STDERR disappeared each time I reduced the number of tests.

I finally published but I had to severely reduce the tests.

**Comment by g964** [comment]
copy of: <https://github.com/Codewars/codewars.com/issues/830>

Sorry to pollute with a long post... I translated to swift <https://www.codewars.com/kata/5726f813c8dcebf5ed000a6b>
I got lots of
```
Time: 4898ms Passed: 13 Failed: 0
Test Results:
  SolutionTest
  countKprimes (9 Passed, 0 Failed)
  puzzle (4 Passed, 0 Failed)
  Random Examples
STDERR:
0  swift           0x00000000034bf568 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift           0x00000000034c0146
2  libpthread.so.0 0x00007f696139e330
3  libpthread.so.0 0x00007f6961714801
4  libpthread.so.0 0x00007f69617153ff
5  libpthread.so.0 0x00007f6961715d50
6  libpthread.so.0 0x00007f6961714b2e
7  libpthread.so.0 0x00007f6961715e75
8  libXCTest.so    0x00007f695c64e6ab
9  libXCTest.so    0x00007f695c6251a2
10 libXCTest.so    0x00007f695c64e660
11 libXCTest.so    0x00007f695c64e3e0
12 libXCTest.so    0x00007f695c624280 _TFC6XCTest10XCTestCase10invokeTestfT_T_ + 112
13 libXCTest.so    0x00007f695c63c219 _TTSf4g_n___TFC6XCTest10XCTestCase7performfCS_9XCTestRunT_ + 313
14 libXCTest.so    0x00007f695c62414e _TFC6XCTest10XCTestCase7performfCS_9XCTestRunT_ + 14
15 libXCTest.so    0x00007f695c623f45 _TFC6XCTest6XCTest3runfT_T_ + 629
16 libXCTest.so    0x00007f695c63cee5 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 789
17 libXCTest.so    0x00007f695c63d3b3 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 2019
18 libXCTest.so    0x00007f695c63d3b3 _TTSf4g_n___TFC6XCTest11XCTestSuite7performfCS_9XCTestRunT_ + 2019
19 libXCTest.so    0x00007f695c6296db _TF6XCTest7XCTMainFGSaT13testCaseClassMCS_10XCTestCase8allTestsGSaTSSFzS0_T_____Os5Never + 4283
20 libXCTest.so    0x00007f69617140e9 _TF6XCTest7XCTMainFGSaT13testCaseClassMCS_10XCTestCase8allTestsGSaTSSFzS0_T_____Os5Never + 84851401
21 swift           0x0000000000c795f2 llvm::MCJIT::runFunction(llvm::Function*, llvm::ArrayRef) + 1042
22 swift           0x0000000000c7d0ff llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::vector > const&, char const* const*) + 1215
23 swift           0x000000000083ca6f swift::RunImmediately(swift::CompilerInstance&, std::vector > const&, swift::IRGenOptions&, swift::SILOptions const&) + 2303
24 swift           0x00000000008364f8
25 swift           0x0000000000831377 swift::performFrontend(llvm::ArrayRef, char const*, void*, swift::FrontendObserver*) + 2551
26 swift           0x00000000007f8788 main + 3016
27 libc.so.6       0x00007f695fb47f45 __libc_start_main + 245
28 swift           0x00000000007f5e56
Stack dump:
0.	Program arguments: /swift-vol/swift-gen/usr/bin/swift -frontend -interpret /tmp/swift117117-18-p227nr.n1qsuzyqfr/solution.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -module-name solution
````
I don't understand why that:-( Is there a problem with Swift and if yes, which problem? These STDERR disappeared each time I reduced the number of tests.

I finally published but I had to severely reduce the tests.

**Comment by g964** [comment]
There is a bug in Linux-Swift-3 strings comparison, see: <https://bugs.swift.org/browse/SR-530 >, http://stackoverflow.com/questions/42317699/comparing-strings-in-swift-3-gives-false-results-linux. I experienced that problem:-(

**Pull request merged**